Simplified/OPDS/NYPLOPDSFeedFetcher.swift

```diff
@@ -0,0 +1,118 @@
+  ///   - Parameter completion: Always invoked at the end no matter what,
+  ///   providing an ungrouped feed object in case of success and nil otherwise.
+  class func fetchCatalogUngroupedFeed(url: URL,
+                                       retryCount: Int,
+                                       completion: @escaping (_ feed: NYPLCatalogUngroupedFeed?) -> Void) {
+    
+    if (retryCount >= fetchSupportedBooksRetryThreshold) {
+      Log.info(#function, "Retry threshold reached while fetching Catalog feed.")
+      DispatchQueue.global(qos: .userInitiated).async {
+        completion(nil)
+      }
+      return
+    }
+    
+    fetchOPDSFeed(url: url, shouldResetCache: false) { feed, error in
+      guard let feed = feed,
+            feed.type == NYPLOPDSFeedType.acquisitionUngrouped else {
+        DispatchQueue.global(qos: .userInitiated).async {
+          completion(nil)
+        }
+        return
+      }
+    
+      let catalogFeed = NYPLCatalogUngroupedFeed.init(opdsFeed: feed)
+      if let catalogFeed = catalogFeed,
+          catalogFeed.books.count == 0 {
+        fetchCatalogUngroupedFeed(url: catalogFeed.nextURL,
+                                  retryCount: retryCount + 1,
+                                  completion: completion)
+      } else {
+        DispatchQueue.global(qos: .userInitiated).async {
+          completion(catalogFeed)
+        }
+      }
+    }
+  }
+  
+  ///   Fetch OPDS feed with a GET request for the given URL.
+  ///
+  ///   - Parameter url: The URL to contact
+  ///   - Parameter shouldResetCache: Pass YES to wipe the whole cache.
+  ///   - Parameter completion: Always invoked at the end no matter what,
+  ///   providing an OPDS feed object in case of success and an dictionary containing error information otherwise.
+  class func fetchOPDSFeed(url: URL,
+                           shouldResetCache: Bool,
+                           completion: @escaping (_ feed: NYPLOPDSFeed?, _ error: [String: Any]?) -> Void) {
+    let cachePolicy: NSURLRequest.CachePolicy = shouldResetCache ? .reloadIgnoringCacheData : .useProtocolCachePolicy
+    
+    _ = NYPLNetworkExecutor.shared.GET(url,
+                                       cachePolicy: cachePolicy) { result, response, error in
+      
+      if let error = error as NSError? {
+        // Note: NYPLNetworkExecutor already logged this error
+        DispatchQueue.global(qos: .default).async {
+          completion(nil, error.problemDocument?.dictionaryValue)
+        }
+        return
+      }
+      
+      guard let feedXML = NYPLXML.init(data: result) else {
+        Log.info(#function, "Failed to parse data as XML.")
+        NYPLErrorLogger.logError(withCode: .feedParseFail,
+                                 summary: "NYPLOPDSFeed: Failed to parse data as XML",
+                                 metadata: [
+                                  "requestURL": url.absoluteString as Any,
```

Comment:
we have a helper method for URLRequest called `loggableString`, I would just use that instead because it includes the `url` plus more info.

Reply:
That's what I was doing and its what causing the crash I mentioned in the ticket. The `URLRequest` object is released from the memory when it reaches this line of code, giving an error code `EXC_BAD_ACCESS`. Which is really weird since `URLRequest` is passed by value but not reference, and should not cause such error? It happens only on iOS 15 from what I am seeing. Any thoughts?

Reply:
actually now I don't see the urlRequest instance.... maybe I got confused. We can keep the URL.